forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java

```diff
@@ -286,21 +286,34 @@ public VCS repositoryType() {
     }
 
     @Override
-    public String fileContents(String filename, String ref) {
-        var confName = URLEncoder.encode(filename, StandardCharsets.UTF_8);
-        var conf = request.get("repository/files/" + confName)
-                          .param("ref", ref)
-                          .onError(response -> {
-                              log.warning("First time request returned bad status: " + response.statusCode());
-                              log.info("First time response body: " + response.body());
-                              // Retry once with additional escaping of the path fragment
-                              var escapedConfName = URLEncoder.encode(confName, StandardCharsets.UTF_8);
-                              return Optional.of(request.get("repository/files/" + escapedConfName)
-                                            .param("ref", ref).execute());
-                          })
-                          .execute();
-        var content = Base64.getDecoder().decode(conf.get("content").asString());
-        return new String(content, StandardCharsets.UTF_8);
+    public Optional<String> fileContents(String filename, String ref) {
+        var encodedFileName = URLEncoder.encode(filename, StandardCharsets.UTF_8);
+        JSONValue content;
+        try {
+            content = request.get("repository/files/" + encodedFileName)
+                    .param("ref", ref)
+                    .onError(response -> {
+                        // Retry once with additional escaping of the path fragment
+                        // Only retry when the error is exactly "File Not Found"
+                        if (response.statusCode() == 404 && response.body().contains("File Not Found")) {
+                            log.warning("First time request returned bad status: " + response.statusCode());
+                            log.info("First time response body: " + response.body());
+                            var doubleEncodedFileName = URLEncoder.encode(encodedFileName, StandardCharsets.UTF_8);
+                            return Optional.of(request.get("repository/files/" + doubleEncodedFileName)
+                                    .param("ref", ref).execute());
+                        }
+                        return Optional.empty();
+                    })
+                    .execute();
+        } catch (UncheckedRestException e) {
+            // After retry, if the error is still 404 File Not Found, return Optional.empty().
+            if (e.getStatusCode() == 404 && e.getBody().contains("File Not Found")) {
+                return Optional.empty();
+            }
+            throw e;
```

Comment:
Instead of catching exception, I think it's better to register yet another `.onError` handler on the nested request object. You would then return a special JSONValue instance that you can check for and then the outer method can return `Optional.empty()`. There are other examples of this pattern in this file.